### PR TITLE
Handle UPN name constraints more consistently across platforms

### PR DESCRIPTION
**Pull request by vcsjones**
Draft run to get feedback from different chain builders from OSes.

### CODE REVIEW

File: src/libraries/Common/src/System/Security/Cryptography/Asn1/GeneralSubtreeAsn.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      GeneralSubtree ::= SEQUENCE {
+          base                    GeneralName,
+          minimum         [0]     BaseDistance DEFAULT 0,
+          maximum         [1]     BaseDistance OPTIONAL }
+
+      BaseDistance ::= INTEGER (0..MAX)
+  -->
+  <asn:AsnType name="@base" typeName="System.Security.Cryptography.Asn1.GeneralNameAsn" />
+  <asn:Integer name="minimum" backingType="int" explicitTag="0" defaultDerInit="0x02, 0x01, 0x00" />
+  <asn:Integer name="maximum" backingType="int" explicitTag="1" optional="true" />
```
Comment (bartonjs) [inline src/libraries/Common/src/System/Security/Cryptography/Asn1/GeneralSubtreeAsn.xml:19]: I believe we've (I've) generally used PascalCasing (even though that's wrong for fields... were they ever properties?).  It solves the need for "@base", and makes "Defaultminimum" => "DefaultMinimum"

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Apple.cs [hunk 2]
```diff
@@ -305,6 +308,57 @@ private static (X509Certificate2, int)[] ParseResults(
...
+                    return;
+                }
+
+                if (AnyUpnConstraint(asn.excludedSubtrees) || AnyUpnConstraint(asn.permittedSubtrees))
+                {
+                    elements[i] = (cert, flags | (int)X509ChainStatusFlags.HasNotSupportedNameConstraint);
+                }
+            }
+
+            static bool AnyUpnConstraint(GeneralSubtreeAsn[]? subtreeAsns)
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Apple.cs:343]: Should we be opting UPN in as unknown, or say "they only handle 4 kinds, we'll reject anything else"?

File: src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs [hunk 3]
```diff
@@ -594,6 +596,102 @@ public static void NameConstraintViolation_InvalidGeneralNames()
...
         }
 
+        [Fact]
+        public static void NameConstraintViolation_ExcludedTree_Upn()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddUserPrincipalName("mona@github.com");
+
+            AsnWriter writer = new(AsnEncodingRules.DER);
+            writer.WriteCharacterString(UniversalTagNumber.UTF8String, "@github.com");
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs:606]: It's probably more hygenic to use something in a reserved namespace, like `someone@test.example`

File: src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs [hunk 4]
```diff
@@ -594,6 +596,102 @@ public static void NameConstraintViolation_InvalidGeneralNames()
...
+                }
+                else
+                {
+                    Assert.Equal(
+                        PlatformNameConstraints(X509ChainStatusFlags.HasNotSupportedNameConstraint, true),
+                        chain.AllStatusFlags());
+                }
+            });
+        }
+
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs:694]: Consider adding tests with a UPN constraint that passes (on Windows), so we're set up for a future where we gain platform support (or write it ourselves).

Comment (vcsjones) [reply src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs:694]: Yeah, incomplete tests was one of the reason I put a "NO REVIEW" label on the issue :-)

File: src/libraries/Common/src/System/Security/Cryptography/Asn1/GeneralSubtreeAsn.xml (hunk 1 above)
Comment (vcsjones) [reply src/libraries/Common/src/System/Security/Cryptography/Asn1/GeneralSubtreeAsn.xml:19]: I don't think they've ever been properties.

**Comment by vcsjones** [comment]
Closing for now, pending a separate discussion.

**Pull request closed without merging**